ls1intum/Artemis
# `Exam mode`: Collect exam interactions and setup initial monitoring structure

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


As part of my bachelor thesis, we want to get more insight into the exam. Especially as an instructor it is difficult to understand in retrospect how the exam went and especially how the students dealt with it. In order to avoid this problem, we offer the possibility to monitor the exam. This pull request provides the basic structure for collecting the user exam interactions on the client side and persisting them on the server side.

### Description

Each action is collected on the client and stored as part of the `StudentExam`.
To synchronize the actions with the server, we use the same logic which we use to submit and synchronize the exercises/submissions.

- If the server is unreachable, the client stores the actions in local storage and synchronize them when the client is connected again.
- If the server is reachable, the client performs a `put` request (`/api/courses/{courseId}/exams/{examId}/student-exams/{studentExamId}/actions`)
  - if the request was successful, the client deletes the stored actions (no longer needed)
  - if the request was not successful, the actions are still available on the client and not removed

Since Artemis uses multiple instances, we need to share the collected actions and activities between them. For this purpose we use the distributed cache [Hazelcast](https://hazelcast.com/). During the exam, we don't want to store the actions in the database to avoid any conflicts (e.g. performance issues, ...). Therefore we save the activities and actions in the cache during the exam (and save them in the database after the exam -> future PR). 

Some comments to the current progress:

- The collection of the actions on the client needs some fine-tuning (this will be done in an additional PR)

### Steps for Testing

Testing this pull request is a little more complex.

Prerequisites:
- 1 Admin
- 1 or multiple Students
- 1 Exam with multiple Exercises
- Test Server + Test server with multiple Artemis instances (test both)

1. Log in to Artemis
2. Create an exam as admin
3. Enable the exam monitoring (Tick the `Enable Exam Monitoring` checkbox)
4. Create multiple exercises (regardless of the type of exercise)
5. Register one or multiple Students for the exam
6. Create the student exams (Two options -> need to be tested in separate exams with monitoring enabled)
a) Increase the working time of the student
b) Keep the working time as it is -> the exam end is the same as the student exam end
7. Participate as a student
8. Switch to the admin/instructor
9. Verify stored actions in Cache


### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage



| Class/File | Branch | Line |
|------------|-------:|-----:|
| exam-monitoring.service.ts | 85.71% | 100% |
| ExamMonitoringScheduleService.java | 62% | 91% |
| ExamCache.java | n/a | 62% |
| ExamMonitoringDistributedCache.java | 100% | 97% |
| ExamActivityResource.java | 100% | 100% | 
| ExamActivitySaveTask.java | n/a | 100% |
| SavedExerciseAction.java | n/a | 100% |
| StartedExamAction.java | n/a | 100% |
| SwitchedExerciseAction.java | n/a | 100% |
| ConnectionUpdatedAction.java | n/a | 100% |
| HandedInEarlyAction.java | n/a | 100% |
| ContinuedAfterHandedInEarlyAction.java | n/a | 100% |
| EndedExamAction.java | n/a | 100% |


### Screenshots



<details>
<summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/60005702/167831683-49bb12a9-c21d-4a3b-8ff6-18b62199a4c5.png)
</details>
